http2: implement ref() and unref() on sessions

kjin opened the pull request:
Fixes #16617

This is my stab at adding `ref` and `unref` on `ClientHttp2Session` instances. It seems mostly straightforward, but there were a few questions I had in mind about this code change:

- Should `ClientHttp2Session` instances be `unref`d by default?
- Should these be available on `ServerHttp2Session` objects as well? (my thought is no)
- Should calling `ref`/`unref` be a no-op after the session has been destroyed, or could it return an error?

I saw [this comment](https://github.com/nodejs/node/issues/16617#issuecomment-340571382) made by @addaleax about `uv_prepare_t`, but I don't really understand it. To me, this code change seems sufficient, but maybe I'm missing something.

Also, I updated the docs to separate `ClientHttp2Session` from `Http2Session`... which seemed to make sense considering that this distinction is present even in the codebase.

##### Checklist
- [ ] `make -j4 test` (UNIX), or `vcbuild test` (Windows) passes
- [x] tests and/or benchmarks are included
- [x] documentation is changed or added
- [x] commit message follows [commit guidelines](https://github.com/nodejs/node/blob/master/CONTRIBUTING.md#commit-message-guidelines)
##### Affected core subsystem(s)
http2

addaleax commented on doc/api/http2.md:
```diff
@@ -572,6 +515,80 @@ The `http2session.type` will be equal to
+
+#### clienthttp2session.unref()
+<!-- YAML
+added: ??? (TODO)
```
Can you use `added: REPLACEME` here? That way it gets picked up automatically by the release tooling :)

addaleax commented on lib/internal/http2/core.js:
```diff
@@ -1086,6 +1086,18 @@ class ClientHttp2Session extends Http2Session {
   }
+
+  ref() {
+    if (this[kSocket]) {
```
Can you also check whether `typeof this[kSocket].ref === 'function'`? HTTP/2 supports any stream as the underlying socket, so `.ref` might not be present at that point.

The test file that checks that is `test/parallel/test-http2-generic-streams.js`, if you want to extend that – I think calling `.ref()` there would crash right now?

kjin commented:
> > Should these be available on `ServerHttp2Session` objects as well? (my thought is no)
>
> Why not? :)

I thought `http.Server` instances didn't have them, but I now see that they inherit `ref` and `unref` from `net.Server`. I suppose I couldn't think of use cases for `unref`ing an `http.Server` earlier, but there's no reason now to not include them on `Http2Session`.

Seems like it's worthwhile to ensure `unref` works on `Http2Server` as well, so doing both now.

addaleax commented:
@kjin I think those are different things – `ServerHttp2Session` corresponds to a single connection, `Http2Server` to the entire server, right? I’m not sure but they probably both have ref/unref methods?

kjin commented:
Yep, I understand that `ServerHttp2Session` objects are created on a per-connection basis by an `Http2Server` instance, and isn't exactly an analogue of `http.Server`. I definitely skipped an explanation step there but what I was intending to say was "no reason why server-side objects shouldn't also have `ref` and `unref`".

Currently `ServerHttp2Session` doesn't have `unref`, but `Http2Server` does. I think the latter doesn't do anything, and what I was saying above was that I might as well fix that now (whatever that entails).

kjin commented:
@addaleax Thanks for the feedback so far, I've updated the PR to make `ref` and `unref` properties of `Http2Session` instead of `ClientHttp2Session` (and the docs accordingly). I kept the docs for `ClientHttp2Session#request` separate for the same reason as in the initial PR description. Can you take another look when you have the chance?

jasnell commented:
This should wait to land until after https://github.com/nodejs/node/pull/17406, as is may need to be updated based on those changes.

TimothyGu commented:
@addaleax 

> > Should ClientHttp2Session instances be unrefd by default?
>
> I don’t think so; if you have an active network connection, you usually want that to keep the event loop alive.

Hmm, but if the session isn't transporting any data (just kept open because of HTTP/2 semantics) I don't see a good reason why it should keep the event loop alive.

addaleax commented on lib/internal/http2/core.js:
```diff
@@ -997,6 +997,24 @@ class Http2Session extends EventEmitter {
+  ref() {
+    if (this[kSocket]) {
+      assert(this[kSocket]._handle);
+      if (typeof this[kSocket]._handle.ref === 'function') {
```
Why not just `this[kSocket].ref`? That’s what’s being called anyway, right?

kjin replied:
For generic duplex streams `this[kSocket]` is a `JSStreamWrap` object, which doesn't have a `ref`/`unref` field.

Actually, this makes me think that `net.Socket` itself should be checking that `_handle.ref` is a function, rather than here. I'll add a trailing commit that does so, we can drop it if there is already a good rationale for not checking in `net`.

addaleax commented:
@TimothyGu This is just calling through to the underlying socket, the HTTP2 session itself won’t keep anything open

kjin commented:
@jasnell @addaleax I've rebased the PR on master. The only conflict was in the docs, since I moved `http2session.request` to a new `ClientHttp2Session` section. The rebased commits should reflect whatever changed in the docs since I first opened this PR.

I can understand if it's not in the scope of this PR to move the `request` docs, but I still think it makes sense considering that `ClientHttp2Session` is treated as a separate class in `http2/core.js`.

jasnell commented:
CI: https://ci.nodejs.org/job/node-test-pull-request/12269/

jasnell commented:
landed in 120ea9b

ofrobots commented:
@gibfahn This still needs to be released on 8.x. Let myself or @kjin know if there is any work needed from our side to make that happen.

gibfahn commented:
Thanks for the ping, not sure why this hasn't been backported yet.

kjin commented:
@gibfahn This PR was originally contingent on #17406 landing, so I assume that the backport will be as well.
